engine/resource/src/resource_chunk_cache.cpp

```diff
@@ -0,0 +1,445 @@
+// Copyright 2020-2024 The Defold Foundation
+// Copyright 2014-2020 King
+// Copyright 2009-2014 Ragnar Svensson, Christian Murray
+// Licensed under the Defold License version 1.0 (the "License"); you may not use
+// this file except in compliance with the License.
+//
+// You may obtain a copy of the License, together with FAQs at
+// https://www.defold.com/license
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include "resource_chunk_cache.h"
+
+#include <dlib/static_assert.h>
+#include <dlib/array.h>
+#include <dlib/hash.h>
+#include <dlib/log.h>
+
+#include <stdio.h>   // printf
+#include <algorithm> // lower_bound
+
+struct DLListNode
+{
+    struct DLListNode* m_Prev;
+    struct DLListNode* m_Next;
+};
+
+struct DLList
+{
+    DLListNode m_Head;
+    DLListNode m_Tail;
+};
+
+struct ResourceInternalDataChunk
+{
+    DLListNode m_ListNode; // must be first in the struct
+
+    dmhash_t m_PathHash;
+    uint8_t* m_Data;
+    uint32_t m_Size;    // Size of data
+    uint32_t m_Offset;  // Offset into the file
+};
+
+struct ResourceChunkCache
+{
+    // The array is sorted on (path_hash, offset)
+    dmArray<ResourceInternalDataChunk*> m_Chunks;
```

Comment:
Instead of doing hashtables and arrays, I opted for an array of structs, and using lower_bound() to find items in the array.